tests: semaphore tests overhaul

maksimmasalski opened the pull request:
Modify current semaphore tests.
I checked the semaphore tests, and find out many gaps.
Overhaul semaphore tests:
1. Modify some tests
2.Doxygen tags update
3. Update text in zassert messages
4. Remove misprints
5. Test cases names change. Some test cases had a semaphore name in
their name, for example "simple_sem", I removed it from the test
case names. Also some test cases used "sema", some used word "sem".
I decided to make standard short word for a semaphore "sem"

Detailed explanation of the changes:
-test_k_sema_init() -updated name to test_sem_init, updated doxygen
tag, updated zassert text
-test_sem_take_timeout() -updated doxygen tag, added zassert to check
that reset was correct, updated zassert text
-test_sem_take_timeout_fails() -updated doxygen tag, added zassert
to check that reset was correct, updated zassert text
-test_sem_take_timeout_forever() -updated doxygen tag, added zassert
to check that reset was correct, updated zassert text
-test_sem_take_multiple() -updated doxygen tag, modified that test,
added one more thread sem_tid_4, with high priority and added one
more semaphore high_prio_long_sem
-test_simple_sem_from_isr() -updated name to test_sem_give_from_isr,
updated doxygen tag, zassert text fix
-test_simple_sem_from_task() -updated name
to test_sem_give_from_thread, updated doxygen tag, zassert text fix

Tested on qemu_x86, qemu_x86_64, reel_board, and iotdk

andyross commented on tests/kernel/semaphore/semaphore/src/main.c:
```diff
@@ -465,96 +591,226 @@ void test_sem_take_timeout_isr(void)
+ * @see k_sem_count_get(), k_sem_give()
+ */
+void test_k_sem_give_take(void)
+{
```
"give_take" doesn't seem like an informative name for this test.  Pretty much all the tests in this file are giving and taking.

It looks like you're testing the clamping of the semaphore count?  Maybe call it "test_k_sem_count_limit()" or something?

andyross commented on tests/kernel/semaphore/semaphore/src/main.c:
```diff
@@ -465,96 +591,226 @@ void test_sem_take_timeout_isr(void)
-			sem_take_multiple_high_prio_helper,
+	k_thread_create(&sem_tid_3, stack_3, STACK_SIZE,
+			sem_take_multiple_high_prio_long_helper,
 			NULL, NULL, NULL,
```
Can you explain the logic change here?  You're adding another thread and another semaphore (with "long" in their names), but doing similar operations on them.  The original test was just about selecting the correct (highest priority) thread to wake up, I think, so I don't understand the purpose.  What's the behavior being tested?  And what does "long" mean?

chen-png replied:
sem_tid_3 and sem_tid_4 are the same highest priority, but the waiting time of sem_tid_3 is longer than sem_tid_4.
if some threads are the same priority, the sem given operation should be decided according to waiting time.
so i want to test that if a sem is available, it should be given to the highest priority and longest waiting thread.

andyross replied:
I meant: can you explain it in the code? :)

maksimmasalski replied:
I will comment in code his idea)

andyross reviewed (dismissed):
Some notes.  Nothing looks "wrong", exactly, but I really don't understand what half of the new stuff is supposed to do.

Can you start by splitting this patch?  Non-logic things like renamed tests, whitespace changes, comment reflows and added ztest logging and failure result strings should be in a separate patch, and each significant change to test logic should either be obviously separable (either in a separate patch or because it's obvious in code) so it can be understood in isolation.

maksimmasalski commented on tests/kernel/semaphore/semaphore/src/main.c:
```diff
@@ -888,24 +1145,48 @@ void test_sem_multi_take_timeout_diff_sem(void)
+			NULL, K_PRIO_PREEMPT(-1), 0,
+			K_NO_WAIT);
+
+	sem_queue_mutual_exclusion(NULL, NULL, NULL);
```
I don't know how to split that patch correctly, because I made it in a single commit. Seems will be necessary to remove all non-logic things changes, commit only with changes for the tests, then add changes of the non-logic things, make second commit...

That test test_sem_queue_mutual_exclusion I wanted to make sure that kernel can provide a counting semaphore abstraction for queuing, and also for mutual exclusion. Seems that I did that wrong, will rework that test.

maksimmasalski replied:
I don't know how to split that patch correctly, because I made it in a single commit. Seems will be necessary to remove all non-logic things changes, commit only with changes for the tests, then add changes of the non-logic things, make second commit...

andyross replied:
I don't want to enforce busywork for busywork's sake, but to be honest: patch splitting/squashing and rebasing is one of those git tricks that you're going to need to learn at some point anyway. :)

Basically, make this patch your HEAD, "`git reset HEAD^`", then "`git add -p`" and answer the questions as they come up as to whether or not you want to add the current chunk to the index.  Say "y" to all the ones you want to save in a single commit, then "`git commit`" to create a new commit with just the index.  Then repeat.

maksimmasalski replied:
Ok, thanks for that tip! I didn't know about that.

maksimmasalski replied:
@andyross I reworked mutual exclusion test, please take a look, then I will start to split changes.

Commit:
tests: semaphore tests overhaul

Modify current semaphore tests.
I checked the semaphore tests, and find out many gaps.
Overhaul semaphore tests:
1. Modify some tests
2.Doxygen tags update
3. Update text in zassert messages
4. Remove misprints
5. Test cases names change. Some test cases had a semaphore name in
their name, for example simple_sem, I removed it from the test
case names. Also some test cases used sema, some used word sem.
I decided to make standard short word for a semaphore sem

Detailed explanation of the changes:
-test_k_sema_init() -updated name to test_sem_init, updated doxygen
tag, updated zassert text
-test_sem_take_timeout() -updated doxygen tag, added zassert to check
that reset was correct, updated zassert text
-test_sem_take_timeout_fails() -updated doxygen tag, added zassert
to check that reset was correct, updated zassert text
-test_sem_take_timeout_forever() -updated doxygen tag, added zassert
to check that reset was correct, updated zassert text
-test_sem_take_multiple() -updated doxygen tag, modified that test,
added one more thread sem_tid_4, with high priority and added one
more semaphore high_prio_long_sem
-test_simple_sem_from_isr() -updated name to test_sem_give_from_isr,
updated doxygen tag, zassert text fix
-test_simple_sem_from_task() -updated name
to test_sem_give_from_thread, updated doxygen tag, zassert text fix

Tested on qemu_x86, qemu_x86_64, reel_board, and iotdk

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

maksimmasalski commented:
Splitted that patch into 2 PRs:
1. PR with the new created semaphore tests is #25620
2. Current PR only has minor changes of the current semaphore tests.
I think, that way will be easier for maintainer to review my PRs.

maksimmasalski commented:
@nashif @andrewboie @chen-png I'm looking for code review of that PR.

enjiamai reviewed (approved):
It looks good. No other question for this PR.

maksimmasalski commented:
@LeiW000 Please check that semaphore code, I developed it together with Peng in January. It is time to merge it:)

LeiW000 reviewed (approved):
It has been there for long time. All comments have already addressed. I think it can be merged now.

MaureenHelm commented:
@andyross @andrewboie please (re)review

nashif commented:
The main problem with this PR is its review-ability. You are way too many unrelated changes in 1 single commit, this makes it very difficult to review and understand the functional changes vs the cosmetic changes.
Please in the future separate functional changes from cosmetic and cleanup changes into own commits. this will make it much easier to review.
